File: 2024/11/2024-11.py
def main():
    stones = [int(s) for s in open("input.txt").read().split()]
    allstones = set()
    #stones = [0]
    for i in range(25):
        print(i,len(stones))
        new = []
        for s in stones:
            if s == 0:
                new.append(1)
            else:
                digs = str(s)
                l = len(digs)
                if l % 2 == 0:
                    l2 = l//2
                    new.append(int(digs[:l2]))
                    new.append(int(digs[l2:]))
                else:
                    new.append(2024*s)
        stones = new
        allstones |= set(stones)
    #print(len(stones),stones)
    print(len(stones))
    print(len(allstones), len(set(stones)))

main()

File: 2024/11/2024-11.lua
local len = string.len
local band, rshift = bit.band, bit.rshift


function main()
    inp={}
    for n in io.open("input.txt"):lines("*n") do
        inp[#inp+1]=tonumber(n)
    end
    for i = 1,25 do
        new = {}
        for _,s in ipairs(inp) do
            if s == 0 then
                new[#new+1] = 1
            else
                ss = tostring(s)
                ln = len(ss)
                if band(ln,1) == 0 then
                --if ln & 1 == 0 then
                    ln2 = rshift(ln,1)
                    --ln2 = ln >> 1
                    new[#new+1] = tonumber(ss:sub(1,ln2))
                    new[#new+1] = tonumber(ss:sub(ln2+1))
                else
                    new[#new+1] = s * 2024
                end
            end
        end
        inp = new
    end
    print(#inp)
end

main()




